### PR TITLE
Parallelize partition replica migrations

### PR DESCRIPTION
**Pull request by mdogan**
With thousands of partitions and/or with large amount of data (a few TB),
rebalancing partitions after adding or removing a member takes
significant amount of time (up to hours).

By parallelizing migrations, it's possible to reduce this time
to a few mins by utilizing network bandwidth. That way cluster will
stabilize in a shorter time. But there's a trade-off; using
most of the network bandwidth for migration data may increase
the latency of concurrent read/write operations.

In summary:

- Replaced the global partition table version with per-partition
versions.
- Parallelized replica migrations of independent partitions.

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/3676

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1414/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1414/Hazelcast-pr-builder-1414.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

**Comment by pveentjer** [comment]
Having multiple TCP connections could make things less problematic. 

In the current POC there are not isolated connections for particular type of traffic. But we could set up dedicated connections for priority messages or messages without a partitionid.

**Comment by pveentjer** [comment]
How do you control the amount of data that is in flight? Having large amount of data offheap is no problem; but if we suck it onheap we can run into various problem including OOME, major gc's etc.

**Comment by mdogan** [comment]
> How do you control the amount of data that is in flight? Having large amount of data offheap is no problem; but if we suck it onheap we can run into various problem including OOME, major gc's etc.

We don't control the amount of data. There's just a configurable number of parallel migration operation count. But with this enhancement in place, in future it will be possible to change logic for number of parallel migrations.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1760/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1760/Hazelcast-pr-builder-1760.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   AntiEntropyCorrectnessTest.testPartitionData:77->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:221->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:225->HazelcastTestSupport.assertTrueEventually:1344->HazelcastTestSupport.assertTrueEventually:1242->PartitionCorrectnessTestSupport.access$000:66->PartitionCorrectnessTestSupport.assertSizeAndData:265->PartitionCorrectnessTestSupport.assertNoMissingData:299 Partition: 18, replica: 1 data is missing on [127.0.0.1]:5703 - c7cf035a-6d3f-41ae-8b88-1c62d53a2bce
[ERROR]   PartitionReplicaStateCheckerTest.shouldNotBeSafe_whenMigrationTasksScheduled:188 expected:<MIGRATION_LOCAL> but was:<SAFE>
[INFO] 
[ERROR] Tests run: 31168, Failures: 2, Errors: 0, Skipped: 1000
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: docs/design/partitioning/07-parallel-migrations.md [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+With the proposed change, partition table version will be removed and master will update specific partition's version 
+on partition updates. On partition table broadcast, receiving members will do version comparisons for individual 
+partitions. So, they can partially apply, partially ignore the received partition table.
+
+During migrations, migration operations will still carry the most recent completed migrations, but only for the partition
+that migration is belongs to. After applying completed migrations of that partition, migration operation will check 
+the local version against the version in `MigrationInfo`. If they are equal operation will be executed, otherwise it will
+be rejected.
+
+For some cases, we need to check if the whole partition table has changed or not, with a simple comparison. Comparing each
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:74]: New heading for `partition table stamp` would make reading easy.

File: docs/design/partitioning/07-parallel-migrations.md [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+### 1 - Global Version to Partition Level
+
+First of the major changes is the removal of the global partition table update ordering. Instead we will version each 
+partition individually. 
+
+#### 1.1 - Current Versioning Mechanism
+
+Currently, master member increments global version on each partition update. When the partition table update is received 
+from the master, members decide whether to apply or to ignore the update comparing their local version to received 
+version. If local version is less than received version then partition table is applied, otherwise it's ignored.
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:51]: maybe we can reference all-checks-to-start-a-migration-operation documentation, if there is any.

File: docs/design/partitioning/07-parallel-migrations.md [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+    }
+    return HashUtil.MurmurHash3_x64_64(bb, 0, bb.length);
+}
+```  
+
+During promotion commit, partition table stamp will be used the check whether local partition table is the same with 
+the one received with promotion. If yes, promotion commit will be ignored. Otherwise, we'll compare versions of individual
+partitions and skip promotions of partitions which have greater version locally.    
+   
+_Other usages of partition table stamp will be explained in later sections._
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:94]: to see all usages in one place some referencing can help here.

File: docs/design/partitioning/07-parallel-migrations.md [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+
+#### 3.3 - Rolling Upgrade
+
+- To support rolling upgrade, we will have two distinct paths for most the migration scheduling and execution logic. 
+
+- While updating partition table and applying completed migrations, there'll be separate paths too.
+
+- Once the cluster is upgraded to `4.1`, all partition versions will be set to the latest partition table version.
+
+- Cluster state change transaction mechanism will use either partition table version or stamp according to the version.
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:218]: instead of `according to the version`, rewording it as `according to the cluster version` seemed clearer to understand.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1881/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1881/Hazelcast-pr-builder-1881.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MigrationCommitTest.shouldCommitMigrationWhenSourceFailsDuringCommit:301->HazelcastTestSupport.waitAllForSafeState:748->HazelcastTestSupport.waitAllForSafeState:722->HazelcastTestSupport.waitAllForSafeState:726->HazelcastTestSupport.assertTrueEventually:1259->HazelcastTestSupport.assertTrueEventually:1242->HazelcastTestSupport.assertAllInSafeState:763 Instances not in safe state! {[127.0.0.1]:5703=REPLICA_NOT_OWNED, [127.0.0.1]:5701=REPLICA_NOT_OWNED}
[ERROR]   MigrationCommitTest.shouldRollbackMigrationWhenDestinationCrashesBeforeCommit:217->HazelcastTestSupport.waitAllForSafeState:748->HazelcastTestSupport.waitAllForSafeState:722->HazelcastTestSupport.waitAllForSafeState:726->HazelcastTestSupport.assertTrueEventually:1259->HazelcastTestSupport.assertTrueEventually:1242->HazelcastTestSupport.assertAllInSafeState:763 Instances not in safe state! {[127.0.0.1]:5702=MIGRATION_LOCAL, [127.0.0.1]:5701=MIGRATION_LOCAL}
[INFO] 
[ERROR] Tests run: 31168, Failures: 2, Errors: 0, Skipped: 1000
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: docs/design/partitioning/07-parallel-migrations.md [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+
+Third change is in the partial restart decision process. During partial restart, we decide which members to perform the 
+restart, which members to be excluded and delete their data. To select member groups having the same partition table,
+we use the partition table version. Members with the same partition table are grouped together. Also to find the most 
+recent partition table, we pick the partition table with the greatest version. If the policy is `PARTIAL_RECOVERY_MOST_RECENT`,
+we pick the group the greatest partition table version. If it's `PARTIAL_RECOVERY_MOST_COMPLETE`, then we pick 
+the largest group. In other places we will use _partition table stamp_ instead of removed partition table version. But in
+this case, we cannot use _partition table stamp_. Because stamp is only valid for equality comparison, it cannot be used
+to compare order of the partition tables. Instead we will use the sum of per-partition versions to compare the order. 
+_This works because Hot Restart does not support arbitrary server crashes. Only proper server shutdowns are supported._
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:204]: `This works because Hot Restart does not support arbitrary server crashes` according to this note, no regression exists compared to old versioning mechanism, but i wonder if there is any regression for the _effective_ behavior (i mean for undocumented working of the system) of old mechanism with global-partition-table-version? For example, can we say, in some arbitrary server crash scenarios, older mechanism is better?

File: docs/design/partitioning/07-parallel-migrations.md [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+| Mean interval between collections (ms) |	180394 |
+| Proportion of time spent in garbage collection pauses (%) |	3.39 |
+| Proportion of time spent unpaused (%) |	96.61 |
+
+
+---
+
+#### Parallel Migrations - 5x
+
+5x parallelization was used: `hazelcast.partition.max.parallel.migrations=5`.
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:372]: I can see no tests with default value 10.

File: docs/design/partitioning/07-parallel-migrations.md [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+| Proportion of time spent in garbage collection pauses (%)	| 4.27
+| Proportion of time spent unpaused (%)	| 95.73
+
+---
+
+#### Parallel Migrations - 20x
+
+20x parallelization was used: `hazelcast.partition.max.parallel.migrations=20`.
+
+- *Scale down from 10 to 9 members*: `14000` migrations were completed in **3 minutes and 22 seconds**.
```
Comment (ahmetmircik) [inline docs/design/partitioning/07-parallel-migrations.md:414]: i can't calculate the improved percentage compared to current master. Congrats for amazing results. From half an hour to 3 minutes. 👏

File: docs/design/partitioning/07-parallel-migrations.md (hunk 2 above)
Comment (mdogan) [reply docs/design/partitioning/07-parallel-migrations.md:51]: Afaik, we haven't documented all of these verifications before. Even some of them are introduced eventually with bug-fixing.

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClusterViewListenerService.java [hunk 8]
```diff
@@ -57,6 +60,12 @@
     private final boolean advancedNetworkConfigEnabled;
     private final AtomicBoolean pushScheduled = new AtomicBoolean();
     private final CoalescingDelayedTrigger delayedPartitionUpdateTrigger;
+    // This is an emulation of the pre-4.1 partition state version.
+    // We will increment this version if a partition table change is detected
+    // while sending partition table to the client.
+    private final AtomicInteger partitionTableVersion = new AtomicInteger();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClusterViewListenerService.java:66]: it can be better to document when this `partitionTableVersion` can be deleted from code.

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientPartitionServiceImpl.java [hunk 9]
```diff
@@ -106,10 +106,9 @@ private boolean shouldBeApplied(Connection connection, Collection<Map.Entry<UUID
             }
             return true;
         }
-        if (partitionStateVersion <= current.partitionSateVersion) {
+        if (partitionStateVersion == current.partitionSateVersion) {
             if (logger.isFinestEnabled()) {
-                logFailure(connection, partitionStateVersion, current,
-                        "response state version is old");
+                logFailure(connection, partitionStateVersion, current, "response state version is old");
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientPartitionServiceImpl.java:111]:
```suggestion
                logFailure(connection, partitionStateVersion, current, "response partition state version is old");
```

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java [hunk 10]
```diff
@@ -271,17 +272,22 @@ private void validateClusterVersionChange(Version newClusterVersion) {
...
-    private void checkMigrationsAndPartitionStateVersion(ClusterStateChange stateChange, int partitionStateVersion) {
-        final InternalPartitionService partitionService = node.getPartitionService();
-        final int thisPartitionStateVersion = partitionService.getPartitionStateVersion();
+    private void checkMigrationsAndPartitionStateStamp(ClusterStateChange stateChange, long partitionStateStamp) {
+        InternalPartitionService partitionService = node.getPartitionService();
+        long thisPartitionStateStamp;
+        if (clusterVersion.isGreaterOrEqual(Versions.V4_1)) {
+            thisPartitionStateStamp = partitionService.getPartitionStateStamp();
+        } else {
+            thisPartitionStateStamp = partitionService.getPartitionStateVersion();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java:281]: can't we remove `partitionStateStamp` from code-base and continue to use `partitionStateVersion` instead? It also helps hot-restart partial restart case.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/InternalPartitionService.java [hunk 11]
```diff
@@ -79,6 +79,11 @@
      */
     void memberRemoved(Member member);
 
+    /**
+     * Returns whether partition table is initialized or not.
+     */
+    boolean isPartitionStateInitialized();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/InternalPartitionService.java:85]:
```suggestion
    boolean isPartitionTableInitialized();
```

File: hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionRuntimeState.java [hunk 12]
```diff
@@ -34,27 +38,42 @@
 import static com.hazelcast.internal.serialization.impl.SerializationUtil.writeNullableCollection;
 import static com.hazelcast.internal.util.StringUtil.LINE_SEPARATOR;
 
-public final class PartitionRuntimeState implements IdentifiedDataSerializable {
+public final class PartitionRuntimeState implements IdentifiedDataSerializable, Versioned {
 
-    private PartitionReplica[] replicas;
-    private int[][] minimizedPartitionTable;
+    private PartitionReplica[] allReplicas;
+    private int[][] encodedPartitionTable;
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/PartitionRuntimeState.java:44]: can we have a short javaDoc for what is `encodedPartitionTable`?

File: docs/design/partitioning/07-parallel-migrations.md (hunk 5 above)
Comment (mdogan) [reply docs/design/partitioning/07-parallel-migrations.md:204]: I think no, there won't be any observable behavior change.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java [hunk 13]
```diff
@@ -180,6 +195,46 @@ private CallStatus beforePromotion() {
...
+        InternalPartitionServiceImpl partitionService = getService();
+        PartitionStateManager stateManager = partitionService.getPartitionStateManager();
+        Iterator<MigrationInfo> iter = promotions.iterator();
+        while (iter.hasNext()) {
+            MigrationInfo promotion = iter.next();
+            InternalPartitionImpl partition = stateManager.getPartitionImpl(promotion.getPartitionId());
+
+            if (partition.version() >= promotion.getFinalPartitionVersion()) {
+                // ? we can't assume this, since partition might be further updated...
+//                if (promotion.getDestination().equals(partition.getOwnerReplicaOrNull())) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java:221]: are these commented lines leftover?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java [hunk 14]
```diff
@@ -283,6 +283,7 @@ final void initInvocationTarget() throws Exception {
             if (previousTargetMember != null) {
                 // If a target member was found earlier but current target member is null
                 // then it means a member left.
+                // TODO: not always true ??
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java:286]: TODO?

File: docs/design/partitioning/07-parallel-migrations.md (hunk 6 above)
Comment (mdogan) [reply docs/design/partitioning/07-parallel-migrations.md:372]: Initially I was considering 5 as the default value and during PoC (which you can see the results on wiki) I did tests with 5, 10 and 20. But for final tests I just used 5 and 20 to show there's no very significant difference between them. Then I picked 10 as the default value, so a middle-ground.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java [hunk 15]
```diff
@@ -214,16 +225,20 @@ void setInitialState(PartitionTableView partitionTable) {
...
+            InternalPartition newPartition = partitionTable.getPartition(partitionId);
+            if (!foundReplica && newPartition != null) {
                 for (int i = 0; i < InternalPartition.MAX_REPLICA_COUNT; i++) {
-                    foundReplica |= replicas[i] != null;
+                    foundReplica |= newPartition.getReplica(i) != null;
                 }
             }
             partition.reset(localReplica);
-            partition.setInitialReplicas(replicas);
+//            partition.setInitialReplicas(replicas);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java:235]: leftover?

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClusterViewListenerService.java (hunk 8 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/client/impl/ClusterViewListenerService.java:66]: Most probably it will stay until the next major version of the client protocol. I'll add this.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java (hunk 10 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java:281]: Problem is `partitionStateVersion` indicates a monotonic order between different partition states. But with the new mechanism, there's no such an order, members may receive partition updates in different order but still can have the same version. `partitionStateStamp` tries to solve this problem.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/InternalPartitionService.java (hunk 11 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/InternalPartitionService.java:85]: I'll remove this method, as it's not used apart from a test.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java (hunk 13 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java:221]: yes 😃, I'll remove them.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java (hunk 14 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java:286]: hmm this is a left-over comment from a CP testing, not relevant for this PR. but still the comment is true...

**Review by ahmetmircik — APPROVED** [review]
+ minor comments.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java [hunk 16]
```diff
@@ -947,17 +1048,529 @@ public void migrate(PartitionReplica source, int sourceCurrentReplicaIndex, int
...
+                Operation op = new MigrationRequestOperation(migration, completedMigrations, 0, fragmentedMigrationEnabled);
+                future = nodeEngine.getOperationService()
+                        .createInvocationBuilder(SERVICE_NAME, op, fromMember.getAddress())
+                        .setCallTimeout(partitionMigrationTimeout)
+                        .invoke();
+            } catch (Throwable t) {
+                Level level = migration.isValid() ? Level.WARNING : Level.FINE;
+                logger.log(level, "Error during " + migration, t);
+                future = new CompletableFuture<>();
+                future.completeExceptionally(t);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java:1379]: minor: could also use the static factory methods from `InternalCompletableFuture` for completed futures

File: docs/design/partitioning/07-parallel-migrations.md [hunk 17]
```diff
@@ -0,0 +1,536 @@
...
+to reach the target partition table. These `MigrationTask`s are executed serially by migration thread and they submit 
+migration operations to the migration participants. After each migration completion, migration is committed to the 
+destination, and partition table is updated. All of these steps are executed serially.
+
+ 
+#### 2.2 - Parallel Migrations Mechanism
+
+With parallel migrations, most the mechanism remains the same. `RepartitioningTask` prepares the migration plan using 
+the target partition table. Instead of individual `MigrationTask`s, a single `MigrationPlanTask` will be scheduled. 
+`MigrationPlanTask` will submit parallel migration tasks and keep track of started/completed tasks.
```
[the remark this answers is not in the document]
Comment (mdogan) [reply docs/design/partitioning/07-parallel-migrations.md:114]: For the record; `AsyncMigrationTask`s are also executed in `MigrationThread`. They invoke `MigrationRequestOperation` asynchronously, don't wait for the completion. Once migration is committed/finalized, a new `AsyncMigrationTask` is executed.

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientPartitionServiceImpl.java [hunk 18]
```diff
@@ -106,10 +106,9 @@ private boolean shouldBeApplied(Connection connection, Collection<Map.Entry<UUID
             }
             return true;
         }
-        if (partitionStateVersion <= current.partitionSateVersion) {
+        if (partitionStateVersion == current.partitionSateVersion) {
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientPartitionServiceImpl.java:109]: hmm.. good catch! I think I should revert this change back. I was planning to send partition stamp to the clients but couldn't find a way because of client compatibility requirement.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/IPartition.java [hunk 19]
```diff
@@ -98,4 +98,13 @@
      * @return {@code true} if address is owner or backup, {@code false} otherwise
      */
     boolean isOwnerOrBackup(Address address);
+
+    /**
+     * Returns the version of the partition.
+     * Partition version is incremented by one on each update
+     * on the partition.
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/IPartition.java:105]: 👍 I'll elaborate more.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationStats.java [hunk 20]
```diff
@@ -208,11 +208,10 @@ public String formatToString(boolean detailed) {
             s.append(", elapsedMigrationOperationTime=").append(getElapsedMigrationOperationTime()).append("ms")
                     .append(", totalElapsedMigrationOperationTime=").append(getTotalElapsedMigrationOperationTime()).append("ms")
                     .append(", elapsedDestinationCommitTime=").append(getElapsedDestinationCommitTime()).append("ms")
-                    .append(", totalElapsedDestinationCommitTime=").append(getTotalElapsedDestinationCommitTime()).append("ms");
+                    .append(", totalElapsedDestinationCommitTime=").append(getTotalElapsedDestinationCommitTime()).append("ms")
+                    .append(", elapsedMigrationTime=").append(getElapsedMigrationTime()).append("ms")
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationStats.java:212]: These numbers show the total elapsed time of the individual migrations. But when migrations are concurrent, they might be misleading. Because they don't show the absolute elapsed time between start and complete. Still I can revert this change, if you think it's useful...

File: hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationRequestOperation.java [hunk 21]
```diff
@@ -277,13 +264,7 @@ private void logThrowable(Throwable t) {
...
             throwableToLog = throwableToLog.getCause() != null ? throwableToLog.getCause() : throwableToLog;
         }
-        Level level = getLogLevel(throwableToLog);
-        getLogger().log(level, throwableToLog.getMessage(), throwableToLog);
-    }
-
-    private Level getLogLevel(Throwable e) {
-        return (e instanceof MemberLeftException || e instanceof InterruptedException)
-                || !getNodeEngine().isRunning() ? Level.INFO : Level.WARNING;
+        getLogger().warning("Failure while executing " + migrationInfo, throwableToLog);
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationRequestOperation.java:267]: It was for analyzing a failure. I'll revert back.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java [hunk 22]
```diff
@@ -283,6 +283,7 @@ final void initInvocationTarget() throws Exception {
             if (previousTargetMember != null) {
                 // If a target member was found earlier but current target member is null
                 // then it means a member left.
+                // TODO: Above comment is not always true for RaftInvocation.
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java:286]: I'll create an issue. 👍

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationStats.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationStats.java:212]: oh right, so with concurrent migrations it will be more like "cpu time" rather than wall clock time. Let's leave this only in `detailed == true` case since it can be confusing.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java [hunk 23]
```diff
@@ -423,6 +408,86 @@ private boolean commitMigrationToDestination(MigrationInfo migration) {
...
+
+            MigrationCommitOperation operation = new MigrationCommitOperation(migration, destinationUuid);
+            InvocationFuture<Boolean> future = nodeEngine.getOperationService()
+                    .createInvocationBuilder(SERVICE_NAME, operation, destination.address())
+                    .setTryCount(Integer.MAX_VALUE)
+                    .setCallTimeout(memberHeartbeatTimeoutMillis).invoke();
+
+            final String successResult = "SUCCESS";
+            final String failureResult = "FAIL";
+            final String retryResult = "RETRY";
```
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java:447]: 👍 I was logging the responses while testing, that's why I've used string values. I'll convert to `int`s.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java:286]: https://github.com/hazelcast/hazelcast/issues/17420

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2093/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2093/Hazelcast-pr-builder-2093.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MigrationCorrectnessTest>AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate:82->AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated:98->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:221->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:225->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->PartitionCorrectnessTestSupport.access$000:66->PartitionCorrectnessTestSupport.assertSizeAndData:263->PartitionCorrectnessTestSupport.assertNoMissingData:299 Partition: 51, replica: 1 data is missing on [127.0.0.1]:5706 - 6165312c-01ce-4039-b77a-9d0c166d3333
[INFO] 
[ERROR] Tests run: 31216, Failures: 1, Errors: 0, Skipped: 976
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Review by vbekiaris — APPROVED** [review]

**Comment by Holmistr** [comment]
run-lab-run

**Comment by mdogan** [comment]
Failure in previous run (https://github.com/hazelcast/hazelcast/pull/17197#issuecomment-681878382) is mostly same with this issue: https://github.com/hazelcast/hazelcast/issues/17377. I am able to reproduce it in 669th run.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2098/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2098/Hazelcast-pr-builder-2098.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   AntiEntropyCorrectnessTest.testPartitionData:77->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:221->PartitionCorrectnessTestSupport.assertSizeAndDataEventually:225->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->PartitionCorrectnessTestSupport.access$000:66->PartitionCorrectnessTestSupport.assertSizeAndData:263->PartitionCorrectnessTestSupport.assertNoMissingData:299 Partition: 6, replica: 1 data is missing on [127.0.0.1]:5703 - 74d2d592-726d-46c0-bad4-63e1413a5f6e
[ERROR]   MigrationCommitTest.shouldCommitMigrationWhenSourceFailsDuringCommit:299->HazelcastTestSupport.waitAllForSafeState:754->HazelcastTestSupport.waitAllForSafeState:728->HazelcastTestSupport.waitAllForSafeState:732->HazelcastTestSupport.assertTrueEventually:1265->HazelcastTestSupport.assertTrueEventually:1248->HazelcastTestSupport.assertAllInSafeState:769 Instances not in safe state! {[127.0.0.1]:5703=REPLICA_NOT_OWNED, [127.0.0.1]:5701=REPLICA_NOT_OWNED}
[ERROR] Errors: 
[ERROR]   AdvancedClusterStateTest.changeClusterState_shouldNotFail_whenNonInitiatorMemberDies_duringCommit:320 ? Hazelcast
[INFO] 
[ERROR] Tests run: 31216, Failures: 2, Errors: 1, Skipped: 976
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2101/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2101/Hazelcast-pr-builder-2101.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PartitionAssignmentsCorrectnessTest>AbstractPartitionAssignmentsCorrectnessTest.testPartitionAssignments_whenNodesStartedTerminated:48->AbstractPartitionAssignmentsCorrectnessTest.assertPartitionAssignmentsEventually:76->AbstractPartitionAssignmentsCorrectnessTest.assertPartitionAssignmentsEventually:80->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->PartitionCorrectnessTestSupport.assertPartitionAssignments:210 On Member [127.0.0.1]:5702 - 8e1f5f0f-c294-4b59-b62d-d694867c5661 this, Replica 1 is not found in Partition {ID: 53, Version: 6} [
	0:[127.0.0.1]:5701 - 35fdc937-86b9-427c-8ddc-fd528bc9d70a
]
[INFO] 
[ERROR] Tests run: 31222, Failures: 1, Errors: 0, Skipped: 976
[INFO] 

[ERROR] There are test failures.
</details>
</pre>
